### PR TITLE
add logprocessor to agent

### PR DESCRIPTION
**Pull request by kryalama**
add log processor to agent.

### CODE REVIEW

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/AgentProcessor.java [hunk 1]
```diff
@@ -36,20 +36,23 @@ public AgentProcessor(@Nullable IncludeExclude include,
         return exclude;
     }
 
+
```
Comment (ramthi) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/AgentProcessor.java:39]: NIT:  unnecessary newlines.  Also, there seem to be inconsistency around how we use newline before closing paranthesis in a method.  As far as can tell, there seem to be no newline for most methods.  Can you pl. fix?

**Comment by heyams** [comment]
note: missing copyrights for new classes.

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/ExporterWithLogProcessor.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        config.validate();
+        logProcessor = SpanProcessor.create(config);
+        this.delegate = delegate;
+    }
+
+    @Override
+    public CompletableResultCode export(Collection<SpanData> spans) {
+        // we need to filter attributes before passing on to delegate
+
+            List<SpanData> copy = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (trask) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/ExporterWithLogProcessor.java:30]: let's not reformat the whole class yet, since we have two branches going on right now (`aad` and `main`) so could cause merge conflicts

i've created a [work item](https://msazure.visualstudio.com/One/_workitems/edit/10033056) to reformat all code using same format used in OTel Java projects

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/ProcessorUtil.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package com.microsoft.applicationinsights.agent.internal.processors;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.sdk.trace.data.SpanData;
+
+public class ProcessorUtil {
+    private static final AttributeKey<Boolean> AI_LOG_KEY = AttributeKey.booleanKey("applicationinsights.internal.log");
+    public static boolean isSpanOfTypeLog(SpanData span) {
+        Boolean isLog = span.getAttributes().get(AI_LOG_KEY);
+        return isLog != null && isLog;
```
[the remark this answers is not in the document]
Comment (kryalama) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/ProcessorUtil.java:10]: return isLog will throw a NullPointerException

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/AgentProcessor.java [hunk 4]
```diff
@@ -61,18 +64,34 @@ public static StrictIncludeExclude create(ProcessorIncludeExclude includeExclude
...
+            if(spanNames.isEmpty() && logNames.isEmpty()) {
+                // check attributes for both spans and logs
+                return this.checkAttributes(span);
             }
+            if(isLog) {
+               if(logNames.isEmpty()) return false;
+               if(!logNames.isEmpty() && !logNames.contains(span.getName())) return false;
+            } else {
+                if(spanNames.isEmpty()) return false;
+                if(!spanNames.isEmpty() && !spanNames.contains(span.getName())) return false;
```
[the remark this answers is not in the document]
Comment (kryalama) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/AgentProcessor.java]: we still have to check if there are any attributes matching

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/ProcessorUtil.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (heyams) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/ProcessorUtil.java:32]: i was thinking about `new Boolean(null)`.. yup, need null check because of the autocast.

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/AttributeProcessor.java [hunk 5]
```diff
@@ -33,6 +54,19 @@ public static AttributeProcessor create(ProcessorConfig config) {
...
+    // this won't be needed once we update to 0.13.0
+    // see https://github.com/open-telemetry/opentelemetry-java/pull/2284
+    public static String getAttribute(Attributes attributes, AttributeKey<String> key) {
+        Object existingValueObj = attributes.get(key);
+        // checking the return type won't be needed once we update to 0.13.0
+        // see https://github.com/open-telemetry/opentelemetry-java/pull/2284
+        if (existingValueObj instanceof String) {
+            return (String) existingValueObj;
+        } else {
+            return null;
```
Comment (heyams) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/AttributeProcessor.java:66]: else block is not needed. Simply `return null;`

File: test/smoke/testApps/TelemetryProcessors/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java [hunk 6]
```diff
@@ -26,6 +27,13 @@ public void doMostBasicTest() throws Exception {
         assertEquals("/TelemetryProcessors/test", rd.getProperties().get("httpPath"));
         assertEquals(4, rd.getProperties().size());
         assertTrue(rd.getSuccess());
+
+        List<MessageData> logs = mockedIngestion.getAllMessageData();
```
Comment (heyams) [inline test/smoke/testApps/TelemetryProcessors/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java:31]: can we add comment for testing log processor?

File: test/smoke/testApps/SystemExit/src/smokeTest/resources/jre.excludes.txt [hunk 7]
```diff
@@ -0,0 +1,2 @@
+azul/zulu-openjdk:11
```
Comment (heyams) [inline test/smoke/testApps/SystemExit/src/smokeTest/resources/jre.excludes.txt:1]: why do we exclude these? for running locally probably?

File: test/fakeIngestion/standalone/src/main/java/com/microsoft/applicationinsights/test/fakeingestion/MockedAppInsightsIngestionServer.java [hunk 8]
```diff
@@ -93,6 +93,19 @@ public int getCountForType(String type) {
         return dataItems;
     }
 
+    public <T extends Domain> List<T> getAllMessageData() {
```
Comment (heyams) [inline test/fakeIngestion/standalone/src/main/java/com/microsoft/applicationinsights/test/fakeingestion/MockedAppInsightsIngestionServer.java:96]: can you use getMessageDataInRequest?

File: test/smoke/testApps/TelemetryProcessors/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java (hunk 6 above)
Comment (trask) [reply test/smoke/testApps/TelemetryProcessors/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java:31]: makes sense, can move the `logger.info` to the controller method so that it happens during the request, instead of during startup

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/AgentProcessor.java [hunk 9]
```diff
@@ -61,14 +84,26 @@ public static StrictIncludeExclude create(ProcessorIncludeExclude includeExclude
...
+                return this.checkAttributes(span);
+            }
+            if (isLog) {
+                if (logNames.isEmpty()) return false;
+                if (!logNames.isEmpty() && !logNames.contains(span.getName())) return false;
+            } else {
+                if (spanNames.isEmpty()) return false;
+                if (!spanNames.isEmpty() && !spanNames.contains(span.getName())) return false;
             }
             return this.checkAttributes(span);
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/AgentProcessor.java:108]: this method is a bit hard to follow, I think this makes it clearer, but is also slightly different, wdyt?
```suggestion
            // check spanNames and logNames
            if (isLog) {
                if (!spanNames.isEmpty()) {
                    return false;
                }
                if (!logNames.isEmpty() && !logNames.contains(span.getName())) {
                    return false;
                }
            } else {
                if (!logNames.isEmpty()) {
                    return false;
                }
                if (!spanNames.isEmpty() && !spanNames.contains(span.getName())) {
                    return false;
                }
            }
            // check attributes
            return this.checkAttributes(span);
```

File: test/smoke/testApps/SystemExit/src/smokeTest/resources/jre.excludes.txt (hunk 7 above)
Comment (kryalama) [reply test/smoke/testApps/SystemExit/src/smokeTest/resources/jre.excludes.txt:1]: will remove this. Yes this is for local testing

File: test/fakeIngestion/standalone/src/main/java/com/microsoft/applicationinsights/test/fakeingestion/MockedAppInsightsIngestionServer.java (hunk 8 above)
Comment (kryalama) [reply test/fakeIngestion/standalone/src/main/java/com/microsoft/applicationinsights/test/fakeingestion/MockedAppInsightsIngestionServer.java:96]: getMessageDataInRequest will filter logs based on ai.operation.id

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/AgentProcessor.java (hunk 9 above)
Comment (kryalama) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/AgentProcessor.java:108]: @trask this logic will not work for cases where user provides both spanNames and logNames in the processor.

Comment (trask) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/AgentProcessor.java:108]: what does it mean if both are included? should that be valid?

File: test/fakeIngestion/standalone/src/main/java/com/microsoft/applicationinsights/test/fakeingestion/MockedAppInsightsIngestionServer.java (hunk 8 above)
Comment (trask) [reply test/fakeIngestion/standalone/src/main/java/com/microsoft/applicationinsights/test/fakeingestion/MockedAppInsightsIngestionServer.java:96]: can you using `getMessageDataInRequest` now that you've moved the logging into the controller?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/wasbootstrap/configuration/Configuration.java [hunk 10]
```diff
@@ -443,8 +442,6 @@ public void validate(ProcessorType processorType, IncludeExclude includeExclude)
                     validAttributeProcessorIncludeExclude(includeExclude);
                     break;
                 case log:
-                    validateLogProcessorIncludeExclude(includeExclude);
-                    break;
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/wasbootstrap/configuration/Configuration.java:447]: I think it makes sense to keep this, even if it only validates non-empty attributes, in order to keep friendly message specific to log processor

File: test/smoke/appServers/global-resources/telemetryprocessors_applicationinsights.json [hunk 11]
```diff
@@ -1,6 +1,15 @@
...
     "preview": {
+      "selfDiagnostics": {
+        "destination": "file+console",
+        "level": "DEBUG",
+        "file": {
+          "path": "applicationinsights.log",
+          "maxSizeMb": 5,
+          "maxHistory": 1
+        }
+      },
```
Comment (trask) [inline test/smoke/appServers/global-resources/telemetryprocessors_applicationinsights.json:12]: was this just for testing? (also selfDiagnostics doesn't go under preview)

File: test/smoke/testApps/TelemetryProcessors/src/main/java/com/microsoft/applicationinsights/smoketestapp/SpringBootApp.java [hunk 12]
```diff
@@ -1,5 +1,7 @@
 package com.microsoft.applicationinsights.smoketestapp;
 
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
```
Comment (trask) [inline test/smoke/testApps/TelemetryProcessors/src/main/java/com/microsoft/applicationinsights/smoketestapp/SpringBootApp.java:4]: looks like can remove these now

File: test/smoke/testApps/TelemetryProcessors/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java [hunk 13]
```diff
@@ -28,10 +28,11 @@ public void doMostBasicTest() throws Exception {
...
         assertTrue(rd.getSuccess());
         // Log processor test
-        List<MessageData> logs = mockedIngestion.getAllMessageData();
+        List<MessageData> logs = mockedIngestion.getMessageDataInRequest();
         for(MessageData log:logs) {
-            if(log.getMessage().equals("This is log from SpringBootApp.")) {
-                assertEquals("testLogAttributeValue", log.getProperties().get("newLogAttribute"));
+            if(log.getProperties().get("LoggerName") != null && log.getProperties().get("LoggerName").equals("smoketestappcontroller")) {
+                System.out.println(log.getMessage());
+                assertEquals("smoketestappcontroller::INFO",log.getMessage());
```
Comment (trask) [inline test/smoke/testApps/TelemetryProcessors/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java]: can simplify this now, since `getMessageDataInRequest` should only return a single log message

File: test/smoke/appServers/global-resources/telemetryprocessors_applicationinsights.json [hunk 14]
```diff
@@ -77,6 +86,16 @@
             }
           ],
           "id": "attributes/extract"
+        },
+        {
+          "type": "log",
+          "name": {
```
Comment (trask) [inline test/smoke/appServers/global-resources/telemetryprocessors_applicationinsights.json:92]: let's change this to `body` since that's what it's called in OTel log data model

File: test/smoke/testApps/TelemetryProcessors/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java [hunk 15]
```diff
@@ -29,12 +29,8 @@ public void doMostBasicTest() throws Exception {
...
         // Log processor test
         List<MessageData> logs = mockedIngestion.getMessageDataInRequest();
-        for(MessageData log:logs) {
-            if(log.getProperties().get("LoggerName") != null && log.getProperties().get("LoggerName").equals("smoketestappcontroller")) {
-                System.out.println(log.getMessage());
-                assertEquals("smoketestappcontroller::INFO",log.getMessage());
-            }
-        }
+        MessageData md1 = logs.get(0);
+        assertEquals("smoketestappcontroller::INFO",md1.getMessage());
```
Comment (trask) [inline test/smoke/testApps/TelemetryProcessors/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootAutoTest.java:33]: 👍

File: test/smoke/appServers/global-resources/telemetryprocessors_applicationinsights.json [hunk 16]
```diff
@@ -89,7 +80,7 @@
         },
         {
           "type": "log",
-          "name": {
+          "body": {
             "fromAttributes": [
               "applicationinsights.internal.logger_name","applicationinsights.internal.log_level"
             ],
```
Comment (trask) [inline test/smoke/appServers/global-resources/telemetryprocessors_applicationinsights.json]: it would be nice to instead test the PII data redaction scenario, but we can do that in follow-up PR later

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/LogProcessor.java [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+            }
+        }
+        List<List<String>> groupNames = getGroupNamesList(toAttributeRules);
+        String separator = config.body.separator != null ? config.body.separator : "";
+        return new LogProcessor(normalizedInclude, normalizedExclude,
+                fromAttributes, toAttributeRulePatterns, groupNames, separator);
+    }
+
+    //fromAttributes represents the attribute keys to pull the values from to generate the new span name.
+    public SpanData processFromAttributes(SpanData log) {
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/processors/LogProcessor.java:83]: probably could have avoided some of the copy paste here by extracting out another common method into ProcessorUtil, but let's do that later in follow-up PR

**Review by trask — APPROVED** [review]
thanks @kryalama! ❤️
